osquery/logger/plugins/aws_log_forwarder.h

```diff
@@ -0,0 +1,301 @@
+      }
+    }
+
+    return stream;
+  }
+
+  /// Consumes the specified log records generating one or more batches
+  BatchList consumeDataAndGenerateBatches(
+      std::vector<std::string>& discarded_records,
+      const std::string& log_type,
+      std::vector<std::string>& log_data) {
+    BatchList batch_list;
+
+    Batch current_batch;
+    std::size_t current_batch_byte_size = 0U;
+
+    for (auto& record : log_data) {
+      // Initialize the line and make sure we are still within protocol limits
+      Status status = appendLogTypeToJson(log_type, record);
+      if (!status.ok()) {
+        // To achieve behavior parity with TLS logger plugin, skip non-JSON
+        // content
+        LOG(ERROR) << name_ << ": The following log record has been discarded "
+                               "because it was not in JSON format: "
+                   << record;
+
+        continue;
+      }
+
+      std::size_t record_size = record.size();
+      if (appendNewlineSeparators()) {
+        ++record_size;
+      }
+
+      if (record_size >= getMaxBytesPerRecord() ||
+          record_size >= getMaxBytesPerBatch()) {
+        discarded_records.push_back(std::move(record));
+        continue;
+      }
+
+      // Complete the current batch if it's full
+      if (current_batch_byte_size + record_size >= getMaxBytesPerBatch() ||
+          (current_batch.size() >= getMaxRecordsPerBatch())) {
+        batch_list.push_back(current_batch);
+
+        current_batch.clear();
+        current_batch_byte_size = 0U;
+      }
+
+      // Initialize and store the new log record
+      auto buffer = Aws::Utils::ByteBuffer(
+          reinterpret_cast<unsigned char*>(&record[0]), record_size);
+
+      if (appendNewlineSeparators()) {
+        buffer[record_size - 1] = '\n';
+      }
+
+      RecordType aws_record;
+      initializeRecord(aws_record, buffer);
+
+      current_batch.emplace_back(std::move(aws_record));
+      current_batch_byte_size += record_size;
+    }
+
+    if (!current_batch.empty()) {
+      batch_list.push_back(current_batch);
+    }
+
+    log_data.clear();
+    return batch_list;
+  }
+
+ protected:
+  /// Sends the specified data in one or more batches, depending on the log size
+  Status send(std::vector<std::string>& log_data,
+              const std::string& log_type) override {
+    // Generate the batches, according to the protocol limits
+    std::vector<std::string> discarded_records;
+
+    auto batch_list =
+        consumeDataAndGenerateBatches(discarded_records, log_type, log_data);
+
+    for (const auto& record : discarded_records) {
+      LOG(ERROR) << name_ << ": The following log record has been discarded "
+                             "because it was too big: "
+                 << record;
+    }
+
+    discarded_records.clear();
+
+    // Send each batch
+    std::size_t error_count = 0;
+    std::stringstream status_output;
+
+    for (auto batch_it = batch_list.begin(); batch_it != batch_list.end();) {
```

Comment:
nit, remove final `;`

Reply:
I don't think I can remove it; what do you mean exactly?